### ISSUE TITLE
Edit form difficult to use by screen reader and low vision users

### ISSUE THREAD

**Issue by selfthinker** [splitbrain/dokuwiki#3450]
Some parts of the edit form (textarea, toolbar, etc) in the [edit view](https://grimm.haz.wiki/animals?do=edit) cannot be used by blind screen reader users in addition to what is listed in #3449.
These have a medium priority.

* [ ] The **textarea has no accessible name**. That means screen reader users won't have the context what it is for. This is especially bad with VoiceOver on iOS (I haven't tested VO on macOS) as it goes straight into reading the content of the textarea without any other announcement.
That fails [4.1.2 Name, Role, Value](https://www.w3.org/TR/WCAG21/#name-role-value).

![](https://user-images.githubusercontent.com/108893/112762534-0ac0c900-8ff8-11eb-96d6-091f27b40152.png)

The textarea should have a `label`. That could be visually hidden and only accessible to screen reader users.
Another way to give form elements an accessible name is to add it via `aria-label` etc.

* [ ] The toolbar dropdowns (headline, smiley and special character dropdowns) have **`aria-hidden=true` but also include buttons**. [`aria-hidden`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-hidden_attribute) should not be used on anything that contains focusable elements as they won't be re-exposed to assistive technology when changing it to 'false' (at least theoretically when an assistive technologies are following the spec).
That fails [4.1.2 Name, Role, Value](https://www.w3.org/TR/WCAG21/#name-role-value).

The `aria-hidden=true` can be removed when the dropdowns are hidden via `display: none` instead. That is enough to hide something from assistive technologies.

* [ ] It is not programmatically determinable and **not announced in screen readers what the toolbar buttons do** when they open a dropdown. Those toolbar buttons have an [`aria-haspopup`](https://www.w3.org/TR/wai-aria/#aria-haspopup) but it is used incorrectly.
That fails [4.1.2 Name, Role, Value](https://www.w3.org/TR/WCAG21/#name-role-value).

The spec says:
> Authors MUST ensure that the role of the element that serves as the container for the popup content is `menu`, `listbox`, `tree`, `grid`, or `dialog`, and that the value of `aria-haspopup` matches the role of the popup container.

There will probably be more to it than that. Unfortunately I don't have any working knowledge of `aria-haspopup`.
(It could be that `aria-haspopup` announces if a dropdown is open or close when it's implemented correctly. If it doesn't, using `aria-expanded` as well or instead could make sense. I'm not sure.)

This video demonstrates the issue that nothing is announced when activating a dropdown toolbar button using NVDA in Firefox. In it I tab to the smiley dropdown and NVDA announces that it has a 'submenu'. But when I click on it nothing audible happens. Tabbing further only goes further down the toolbar buttons.

https://user-images.githubusercontent.com/108893/112764893-7445d500-9002-11eb-8f1a-1c7c4d649fcd.mp4

* [ ] The **draft status message is not announced to screen reader users** when it pops up.
That fails [4.1.3 Status Messages](https://www.w3.org/TR/WCAG21/#status-messages).

This can be fixed with an [`aria-live` region](https://www.w3.org/TR/wai-aria/#aria-live) or a [`role=status`](https://www.w3.org/TR/wai-aria/#status). 

![](https://user-images.githubusercontent.com/108893/111905853-b17bf700-8a45-11eb-8385-3aac02461380.png)

## Other non-fail accessibility issues

These don't strictly fail WCAG but are still accessibility issues. Fixing them is recommended but optional for passing WCAG.

* [ ] The draft status message will most probably not be noticed by someone using a screen magnifier. (But I don't think drawing attention to it is a good idea in this case either. So, there might not be anything to fix.)

* [ ] A good and consistent heading structure helps screen reader users to navigate a web page. The beginning of each main content of each view should start with an `h1`. That is the case on every page except in the edit view. The main content, i.e. the edit area, is not preceded by a heading. That heading could be visually hidden. (That would make it 4 `h1`s on that page. More on that later.)

* [ ] The toolbar buttons only have `title`s. It would be better for people with cognitive and motor impairments and for general usability reasons if the titles were properly visible. It's [fine for the most common screen readers](https://www.powermapper.com/tests/screen-readers/labelling/button-title-img-null-alt/), though.
During testing with voice control (Dragon) I could activate the buttons without a problem but only because I knew what to say. For example, saying "click bold" would click the button for making text bold. In the case of 'bold' that's more obvious, the names of other buttons is less obvious. And I wouldn't know how to choose smileys etc that way, although there is a fallback that always works but just takes a bit longer ('mouse grid').

* [ ] Screen readers don't read a lot of the special characters properly. For example, NVDA would read the first 22 special character buttons only as "A". But that's a drawback of screen readers and their verbosity levels, so not something that can be easily fixed and only something to be aware of.

* [ ] In High Contrast Mode some toolbar icons are barely readable because they are black (ish) on black. Images should ideally not have a transparent background. Or if they have, they should have a border that contrasts well.
![](https://user-images.githubusercontent.com/108893/112764834-3648b100-9002-11eb-8edc-a0659000dc7d.png)

* [ ] When resizing text in Firefox with "zoom text only", the toolbar icons don't get bigger.
![](https://user-images.githubusercontent.com/108893/112763054-cb47ac00-8ffa-11eb-8fa5-7d75501743ea.png)

* [ ] The toolbar icons get very fuzzy when magnifying the screen.

The toolbar icons are quite old. It might just be time to switch to SVGs. The latter three of the non-fail issues might get fixed by that.

**Comment by doubletaco** [comment]
Was going to report the toolbar keyboard/screen reader issues. As a visually impaired user, any toolbar item with a submenu is completely inaccessible and unusable to me. Core functionality as well as popular and useful plugins like Wrap hook into the toolbar, so they are all equally inaccessible and unusable as a result of the submenu problem.
